src/abstractdataframe/join.jl

```diff
@@ -91,6 +91,346 @@ _rename_cols(old_names::AbstractVector{Symbol},
            (renamecols isa Function ? Symbol(renamecols(string(n))) : Symbol(n, renamecols))
            for n in old_names]
 
+prepare_on_col() = throw(ArgumentError("at least one on column required when joining"))
+prepare_on_col(c::AbstractVector) = c
+prepare_on_col(cs::AbstractVector...) = tuple.(cs...)
+
+# Return if it is allowed to use refpool instead of the original array for joining.
+# There are multiple conditions that must be met to allow for this.
+# In particular we must be able to find a sentinel value that will be used
+# in mapping to signal that the mapping to the longer column is not present
+# in some rare cases it is impossible to find such a sentinel (e.g. for
+# CategoricalArray with missing and having number of levels exactly equal
+# to the typemax of element type of ref pool)
+function check_mapping_allowed(short, long)
+    isempty(short) && return false, nothing
+    isnothing(DataAPI.refpool(long)) && return false, nothing
+    isnothing(DataAPI.invrefpool(long)) && return false, nothing
+
+    T = typeof(DataAPI.refarray(long))
+    T isa Union{Signed, Unsigned} || return false, nothing
+    sentinel = zero(T)
+    haskey(DataAPI.invrefpool(long), sentinel) || return true, sentinel
```

Comment:
Maybe store the result of `DataAPI.invrefpool(long)` given that there's no guaranty that computing it is cheap?

Also, another approach would be to use `first(axes(refarray(long), 1)) + 1` and so on. Not sure whether it has advantages or not.

Reply:
we do not have the guarantee now that `first(axes(refarray(long), 1))` is an integer now in DataAPI.jl. Also doing `+1` could overflow.

Reply:
> Maybe store the result of `DataAPI.invrefpool(long)` given that there's no guaranty that computing it is cheap?

I was thinking about it but I assumed it is cheap. I will change the code.